From 85302b75362df30270383e3a373e60e81b1b2384 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?P=C3=A1draig=20Brady?= <P@draigBrady.com>
Date: Mon, 24 Mar 2014 14:50:23 +0000
Subject: [PATCH] remove runtime dependency on pbr

---
 neutron/version.py | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/neutron/version.py b/neutron/version.py
index 2dbf54d..3c9d912 100644
--- a/neutron/version.py
+++ b/neutron/version.py
@@ -12,6 +12,15 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
-import pbr.version
+class VersionInfo(object):
+    release = "RPMRELEASE"
+    version = "RPMVERSION"
 
-version_info = pbr.version.VersionInfo('neutron')
+    def version_string(self):
+        return self.version
+
+    def release_string(self):
+        return self.release
+
+
+version_info = VersionInfo()
